src/Uno.Core.Tests/Threading/LockAsyncFixture.cs

```diff
@@ -0,0 +1,115 @@
+﻿// ******************************************************************
+// Copyright � 2015-2018 nventive inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// ******************************************************************
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using Uno.Threading;
+
+namespace Uno.Core.Tests.Threading
+{
+	[TestClass]
+	public class LockAsyncFixture
+	{
+		[TestMethod]
+		public async Task BasicLock()
+		{
+			var gate = new AsyncLock();
+			var sync = new object();
+			var gate2 = new object();
+			var current = 0;
+			var maxCurrent = 0;
+			var waiting = 0;
+
+			var func1 = Funcs.Create(async ct => {
+
+				await Task.Yield();
+
+				lock (sync)
+				{
+					waiting++;
+					Monitor.Wait(sync);
+				}
+			
+				using(await gate.LockAsync(ct))
+				{
+					lock (gate2)
+					{
+						current++;
+						maxCurrent = Math.Max(current, maxCurrent);
+					}
+
+					await Task.Delay(100);
+
+					lock (gate2)
+					{
+						current--;
+					}
+				}
+
+				return 42;
+			});
+
+			var c1 = func1(CancellationToken.None);
+			var c2 = func1(CancellationToken.None);
+
+			while (waiting != 2)
+			{
+				await Task.Delay(100);
+			}
+
```

Comment:
If the c1 and c2 takes to long to start, we may face some concurrency issue and get some random failing tests on the build server here ...